### PR TITLE
[SPARK-35768][SQL] Take into account year-month interval fields in cast

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Support take into account year-month interval field in cast

##### Rule cast to target YearMonthIntervalType

|  string  | demo | strict target type  |   months |
|---|---|---|---|
|  [+\|-]y-m | 1-1  | YearMonthIntervalType(YEAR. MONTH) | 13  | 
| [+\|-]y| 1 | YearMonthIntervalType(YEAR. YEAR) | 12  | 
| [+\|-]m | 1 | YearMonthIntervalType(MONTH. MONTH) | 1  | 
|  INTERVAL [+\|-]'[+\|-]y-m' YEAR TO MONTH | interval '1-1' year to month | YearMonthIntervalType(YEAR. MONTH) | 13  | 
|  INTERVAL [+\|-]'[+\|-]m' MONTH | interval '1' month | YearMonthIntervalType(MONTH. MONTH) |  1 |
|  INTERVAL [+\|-]'[+\|-]y' YEAR | interval '1' year | YearMonthIntervalType(YEAR.YEAR) | 12 |

### Why are the changes needed?
Support take into account year-month interval field in cast

### Does this PR introduce _any_ user-facing change?
user can use `cast(str, YearMonthInterval(YEAR, YEAR))` etc

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @MaxGekk

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 1]
```diff
@@ -623,9 +624,9 @@ class CastSuite extends CastSuiteBase {
       YearMonthIntervalType()), 12)
     checkEvaluation(cast(Literal.create("INTERVAL +'1-0' YEAR TO MONTH"),
       YearMonthIntervalType()), 12)
-    checkEvaluation(cast(Literal.create(" interval +'1-0' YEAR  TO MONTH "),
+    checkEvaluation(cast(Literal.create("INTERVAL +'1-0' YEAR  TO MONTH "),
       YearMonthIntervalType()), 12)
-    checkEvaluation(cast(Literal.create(" -1-0 "), YearMonthIntervalType()), -12)
+    checkEvaluation(cast(Literal.create("-1-0"), YearMonthIntervalType()), -12)
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:629]: Why did you do that? This checks case sensitivity and gaps ...

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: > Why did you do that? This checks case sensitivity and gaps ...

Hmmm forgot this case, change back....

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 2]
```diff
@@ -108,18 +108,22 @@ object IntervalUtils {
...
-          case ("-", "-") => toYMInterval(year, month, 1)
-          case ("-", _) => toYMInterval(year, month, -1)
-          case (_, "-") => toYMInterval(year, month, -1)
-          case (_, _) => toYMInterval(year, month, 1)
+          case ("-", "-") => toYMInterval(year, truncatedMonth(month), 1)
+          case ("-", _) => toYMInterval(year, truncatedMonth(month), -1)
+          case (_, "-") => toYMInterval(year, truncatedMonth(month), -1)
+          case (_, _) => toYMInterval(year, truncatedMonth(month), 1)
         }
       case _ => throw new IllegalArgumentException(
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:128]: Could you change the error message, please. It should reflect `startField` and `endField`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 3]
```diff
@@ -108,18 +108,22 @@ object IntervalUtils {
...
+    def truncatedMonth(month: String) : String = {
+      if (endField == YearMonthIntervalType.YEAR) "0" else month
+    }
+
     input.trimAll().toString match {
-      case yearMonthRegex("-", year, month) => toYMInterval(year, month, -1)
-      case yearMonthRegex(_, year, month) => toYMInterval(year, month, 1)
+      case yearMonthRegex("-", year, month) => toYMInterval(year, truncatedMonth(month), -1)
+      case yearMonthRegex(_, year, month) => toYMInterval(year, truncatedMonth(month), 1)
       case yearMonthLiteralRegex(firstSign, secondSign, year, month) =>
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:121]: The regex
```scala
  private val yearMonthLiteralRegex =
    (s"(?i)^INTERVAL\\s+([+|-])?'$yearMonthPatternString'\\s+YEAR\\s+TO\\s+MONTH$$").r
```
doesn't support other year-month interval types, for instance `INTERVAL MONTH`. @cloud-fan Again, this is a duplicate of SQL parser logic.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 4]
```diff
@@ -651,4 +652,54 @@ class CastSuite extends CastSuiteBase {
...
+    checkEvaluation(cast(Literal.create("1-1"),
+      YearMonthIntervalType(MONTH, MONTH)), 13)
+    checkEvaluation(cast(Literal.create("-1-1"),
+      YearMonthIntervalType(YEAR, YEAR)), -12)
+    checkEvaluation(cast(Literal.create("-1-1"),
+      YearMonthIntervalType(YEAR, MONTH)), -13)
+    checkEvaluation(cast(Literal.create("-1-1"),
+      YearMonthIntervalType(MONTH, MONTH)), -13)
+
+    checkEvaluation(cast(Literal.create("INTERVAL '1-1' YEAR TO MONTH"),
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:670]: Could you rewrite this like a loop over a sequence of
(interval string, target type) -> <expected months>

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 5]
```diff
@@ -651,4 +652,54 @@ class CastSuite extends CastSuiteBase {
         checkEvaluation(cast(cast(interval, StringType), YearMonthIntervalType()), period)
       }
   }
+
+  test("SPARK-35768: Take into account year-month interval fields in cast") {
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:656]: Need to test `INTERVAL '...' YEAR` and `INTERVAL '...' MONTH` as well.

**Comment by MaxGekk** [comment]
@AngersZhuuuu Please, resolve conflicts.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:140]: > Could you change the error message, please. It should reflect `startField` and `endField`.

How about current?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 5 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:656]: Done

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu Please, resolve conflicts.

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 6]
```diff
@@ -103,27 +103,45 @@ object IntervalUtils {
...
+      }
+    }
+
     input.trimAll().toString match {
-      case yearMonthRegex("-", year, month) => toYMInterval(year, month, -1)
-      case yearMonthRegex(_, year, month) => toYMInterval(year, month, 1)
+      case yearMonthRegex("-", year, month) => toYMInterval(year, truncatedMonth(month), -1)
+      case yearMonthRegex(_, year, month) => toYMInterval(year, truncatedMonth(month), 1)
+      case yearYearLiteralRegex(firstSign, secondSign, year, _) =>
+        toYMInterval(year, "0", getSigh(firstSign, secondSign))
```
Comment (AngersZhuuuu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:135]: @MaxGekk  Is the logic right here?

**Comment by cloud-fan** [comment]
Can we define the behavior first and put it in the PR description? I'd expect cast string to interval year is equal to: cast string to interval year to month and truncate the month part. No new regex is needed.

**Comment by AngersZhuuuu** [comment]
> Can we define the behavior first and put it in the PR description? I'd expect cast string to interval year is equal to: cast string to interval year to month and truncate the month part. No new regex is needed.

Updated, as your comment, we  don't need to support  `INTERVAL [+|-]'[+|-]y-m' YEAR` `INTERVAL [+|-]'[+|-]y-m' MONTH`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 5 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:656]: @AngersZhuuuu Have your committed the changes? I don't see inputs like `INTERVAL '123' YEAR`

Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:656]: > @AngersZhuuuu Have your committed the changes? I don't see inputs like `INTERVAL '123' YEAR`

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 7]
```diff
@@ -103,27 +103,45 @@ object IntervalUtils {
 
   private val yearMonthPatternString = "([+|-])?(\\d+)-(\\d+)"
   private val yearMonthRegex = (s"^$yearMonthPatternString$$").r
+  private val yearYearLiteralRegex =
+    (s"(?i)^INTERVAL\\s+([+|-])?'$yearMonthPatternString'\\s+YEAR$$").r
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:107]: how about `interval '1' year`?

**Comment by AngersZhuuuu** [comment]
ping @MaxGekk  @cloud-fan after check with PGSQL update the rule in the pr desc

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 8]
```diff
@@ -654,6 +654,33 @@ class CastSuite extends CastSuiteBase {
...
+      ("1-1", YearMonthIntervalType(YEAR, MONTH), 13, 12, 13),
+      ("1-1", YearMonthIntervalType(MONTH, MONTH), 13, 12, 13),
+      ("-1-1", YearMonthIntervalType(YEAR, YEAR), -12, -12, -12),
+      ("-1-1", YearMonthIntervalType(YEAR, MONTH), -13, -12, -13),
+      ("-1-1", YearMonthIntervalType(MONTH, MONTH), -13, -12, -13))
+      .foreach { case (str, dataType, ym, year, month) =>
+        checkEvaluation(cast(Literal.create(str), dataType), ym)
+        checkEvaluation(cast(Literal.create(s"INTERVAL '$str' YEAR TO MONTH"), dataType), ym)
+        checkEvaluation(cast(Literal.create(s"INTERVAL -'$str' YEAR TO MONTH"), dataType), -ym)
+        checkEvaluation(cast(Literal.create(s"INTERVAL '$str' YEAR"), dataType), year)
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:668]: The input interval string `INTERVAL '1-1' YEAR` is incorrect. I guess we must fail here. cc @cloud-fan @srielau

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 9]
```diff
@@ -654,6 +654,33 @@ class CastSuite extends CastSuiteBase {
       }
   }
 
+  test("SPARK-35768: Take into account year-month interval fields in cast") {
+    Seq(("1-1", YearMonthIntervalType(YEAR, YEAR), 12, 12, 12),
+      ("1-1", YearMonthIntervalType(YEAR, MONTH), 13, 12, 13),
+      ("1-1", YearMonthIntervalType(MONTH, MONTH), 13, 12, 13),
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:660]: Use `YearMonthIntervalType(MONTH)` here and in other places.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 9 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 8 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: Done

**Comment by cloud-fan** [comment]
image

Let's follow ANSI first and make it super strict for now: `interval '1-1' year` and `interval '1-1' month` should both fail. We can be more flexible later if there are user requests.

According to ANSI, CAST string to interval should follow the literal rule
image

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:132]: Let's unify them later

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 10]
```diff
@@ -105,25 +105,46 @@ object IntervalUtils {
...
+        case (_, "-") => -1
+        case (_, _) => 1
+      }
+    }
+
     input.trimAll().toString match {
-      case yearMonthRegex("-", year, month) => toYMInterval(year, month, -1)
-      case yearMonthRegex(_, year, month) => toYMInterval(year, month, 1)
+      case yearMonthRegex("-", year, month) => toYMInterval(year, truncatedMonth(month), -1)
+      case yearMonthRegex(_, year, month) => toYMInterval(year, truncatedMonth(month), 1)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:131]: Can we fail here if it not `INTERVAL YEAR TO MONTH`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 11]
```diff
@@ -105,25 +105,46 @@ object IntervalUtils {
...
       case yearMonthLiteralRegex(firstSign, secondSign, year, month) =>
-        (firstSign, secondSign) match {
-          case ("-", "-") => toYMInterval(year, month, 1)
-          case ("-", _) => toYMInterval(year, month, -1)
-          case (_, "-") => toYMInterval(year, month, -1)
-          case (_, _) => toYMInterval(year, month, 1)
+        toYMInterval(year, truncatedMonth(month), getSigh(firstSign, secondSign))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:133]: ditto, let's start with a strict rule and require exact field matching.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 12]
```diff
@@ -105,25 +105,46 @@ object IntervalUtils {
...
+      case yearMonthIndividualLiteralRegex(firstSign, secondSign, value, suffix) =>
+        suffix match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:135]: nit:
```
if ("YEAR".equalsIgnoreCase(suffix)) ... else ...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 13]
```diff
@@ -105,25 +105,46 @@ object IntervalUtils {
...
+          case "YEAR" => toYMInterval(value, "0", getSigh(firstSign, secondSign))
+          case _ =>
+            val months = value.toLong
+            val year = (months / 12).toString
+            val month = (months % 12).toString
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:140]: This looks a bit inefficient. Under the hood, we just need the number of months for YM interval, and it's a waste to calculate year and month and then go back to num months.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 10 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: > Can we fail here if it not `INTERVAL YEAR TO MONTH`?

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 11 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 12 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:138]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 13 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala [hunk 14]
```diff
@@ -1038,6 +1034,13 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
           }.getMessage
           assert(e.contains("Error parsing interval year-month string: integer overflow"))
         }
+
+      Seq("1-1", " 1-8").foreach { interval =>
+          val e = intercept[IllegalArgumentException] {
+            cast(Literal.create(interval), YearMonthIntervalType()).eval()
+          }.getMessage
+          assert(e.contains("Interval string does not match year-month format"))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:1042]: I'm confused. Why should this fail? This is a breaking change

Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:1042]: > I'm confused. Why should this fail? This is a breaking change

Hmmmmmm
image

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:1042]: and we should NOT fail if it's INTERVAL YEAR TO MONTH

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:1042]: `YearMonthIntervalType()` is INTERVAL YEAR TO MONTH

Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala]: Updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 15]
```diff
@@ -105,25 +105,62 @@ object IntervalUtils {
...
 
   def castStringToYMInterval(
       input: UTF8String,
-      // TODO(SPARK-35768): Take into account year-month interval fields in cast
       startField: Byte,
       endField: Byte): Int = {
+
+    val supportedFormat = Map(
+      (YM.YEAR, YM.MONTH) -> Seq("[+|-]y-m", "INTERVAL [+|-]'[+|-]y-m' YEAR TO MONTH"),
+      (YM.YEAR, YM.YEAR) -> Seq("INTERVAL [+|-]'[+|-]y' YEAR"),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:127]: `[+|-]y` should also be fine?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:127]: `cast('1' as interval year)` this works in pgsql

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 16]
```diff
@@ -105,25 +105,62 @@ object IntervalUtils {
...
-          case ("-", _) => toYMInterval(year, month, -1)
-          case (_, "-") => toYMInterval(year, month, -1)
-          case (_, _) => toYMInterval(year, month, 1)
+        checkStringIntervalType(YM.YEAR, YM.MONTH)
+        toYMInterval(year, month, getSign(firstSign, secondSign))
+      case yearMonthIndividualLiteralRegex(firstSign, secondSign, value, suffix) =>
+        val sign = getSign(firstSign, secondSign)
+        if ("YEAR".equalsIgnoreCase(suffix)) {
+          checkStringIntervalType(YM.YEAR, YM.YEAR)
+          sign * Math.toIntExact(value.toLong * MONTHS_PER_YEAR)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:154]: shall we follow `toYMInterval` and add try-catch?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:154]: or we can create a new method to do the try-catch and reuse it.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala [hunk 17]
```diff
@@ -1076,4 +1076,49 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
...
+        checkEvaluation(cast(Literal.create(str), dataType), ym)
+        checkEvaluation(cast(Literal.create(s"INTERVAL '$str' YEAR TO MONTH"), dataType), ym)
+        checkEvaluation(cast(Literal.create(s"INTERVAL -'$str' YEAR TO MONTH"), dataType), -ym)
+      }
+
+    Seq(("13", YearMonthIntervalType(YEAR), 156, 12),
+      ("13", YearMonthIntervalType(MONTH), 156, 13),
+      ("-13", YearMonthIntervalType(YEAR), -156, -12),
+      ("-13", YearMonthIntervalType(MONTH), -156, -13))
+      .foreach { case (str, dataType, year, month) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:1093]: nit:
```
Seq(("13", YearMonthIntervalType(YEAR), 13 * 12),
      ("13", YearMonthIntervalType(MONTH), 13),
      ("-13", YearMonthIntervalType(YEAR), -13 * 12),
      ("-13", YearMonthIntervalType(MONTH), -13))
      .foreach { case (str, dataType, ym) => {
  if (dataType == YearMonthIntervalType(YEAR)) {
          checkEvaluation(cast(Literal.create(s"INTERVAL '$str' YEAR"), dataType), year)
        } else {
          checkEvaluation(cast(Literal.create(s"INTERVAL '$str' MONTH"), dataType), month)
        }
}
```

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:1093]: The value is always number of months, we don't need to check the year and month fields separately.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 15 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:127]: > `cast('1' as interval year)` this works in pgsql

Yes, cast('1' as interval year to month) work well too

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:127]: it's more about consistency: we require extract fields matching and we support both the literal syntax and string payload only.

If "[+|-]y-m", "INTERVAL [+|-]'[+|-]y-m' YEAR TO MONTH" are both supported, I think "[+|-]y", "INTERVAL [+|-]'[+|-]y' YEAR" should be both supported as well.

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: updated

**Review by MaxGekk — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:150]: @AngersZhuuuu Could you open JIRA ticket for this, please.

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:150]: > @AngersZhuuuu Could you open JIRA ticket for this, please.

Sure and will work for ths.

**Pull request closed without merging**